util/conditions/experimental/merge_strategies.go

```
@@ -0,0 +1,408 @@
+// - issues: conditions with positive polarity (normal True) and status False or conditions with negative polarity (normal False) and status True.
+// - unknown: conditions with status unknown.
+// - info: conditions with positive polarity (normal True) and status True or conditions with negative polarity (normal False) and status False.
+func (d *defaultMergeStrategy) Merge(conditions []ConditionWithOwnerInfo, conditionTypes []string, negativeConditionTypes sets.Set[string], stepCounter bool) (status metav1.ConditionStatus, reason, message string, err error) {
+	// Infer which operation is calling this func, so it is possible to use different strategies for computing the message for the target condition.
+	// - When merge should consider a single condition type, we can assume this func is called within an aggregate operation
+	//   (Aggregate should merge the same condition across many objects)
+	isAggregateOperation := len(conditionTypes) == 1
+
+	// - Otherwise we can assume this func is called within an summary operation
+	//   (Aggregate should merge the same condition across many objects)
+	isSummaryOperation := !isAggregateOperation
+
+	// sortConditions the relevance defined by the users (the order of condition types), LastTransition time (older first).
+	sortConditions(conditions, conditionTypes)
+
+	issueConditions, unknownConditions, infoConditions := splitConditionsByPriority(conditions, negativeConditionTypes)
+
+	// Compute the status for the target condition:
+	// Note: This function always return a condition with positive polarity.
+	// - If the top group are issues, use false
+	// - If the top group are unknown, use unknown
+	// - If the top group are info, use true
+	switch {
+	case len(issueConditions) > 0:
+		status = metav1.ConditionFalse
+	case len(unknownConditions) > 0:
+		status = metav1.ConditionUnknown
+	case len(infoConditions) > 0:
+		status = metav1.ConditionTrue
+	}
+
```

What is the fallback / default case here? (I guess it's an error case that should never happen, maybe let's return an error then?)

(same for reason)

---

It does not make sense to have a default when there are no conditions to merge, so the func now returns error in this case.

The default, if it makes sense, is something that the caller must establish.